helix-rest/src/main/java/org/apache/helix/rest/server/resources/metadatastore/MetadataStoreDirectoryAccessor.java

```diff
@@ -111,41 +132,56 @@ public Response deleteMetadataStoreRealm(@PathParam("realm") String realm) {
+   * - "HTTP GET /sharding-keys" which returns all sharding keys in a namespace.
+   * - "HTTP GET /sharding-keys?realm={realm}" which returns sharding keys in the realm.
+   * - "HTTP GET /sharding-keys?prefix={prefix}" which returns sharding keys that have the prefix.
+   * -- JSON response example for this path:
+   * {
+   * 	"prefix": "/sharding/key",
+   * 	"shardingKeys": [{
+   * 		"realm": "testRealm2",
+   * 		"shardingKey": "/sharding/key/1/f"
+   *    }, {
+   * 		"realm": "testRealm2",
+   * 		"shardingKey": "/sharding/key/1/e"
+   *  }, {
+   * 		"realm": "testRealm1",
+   * 		"shardingKey": "/sharding/key/1/b"
+   *  }, {
+   * 		"realm": "testRealm1",
+   * 		"shardingKey": "/sharding/key/1/a"
+   *  }]
+   * }
+   *
+   * - "HTTP GET /sharding-keys?realm={realm}&prefix={prefix}" which returns sharding keys in the
+   * realm and that have the prefix.
    *
-   * @param realm Query param in endpoint path
-   * @return Json representation of a map: shardingKeys -> collection of sharding keys.
+   * @param realm Query param in endpoint path: metadata store realm.
+   * @param prefix Query param in endpoint path: prefix substring of sharding key.
+   * @return Json representation for the sharding keys.
    */
   @GET
   @Path("/sharding-keys")
-  public Response getShardingKeys(@QueryParam("realm") String realm) {
-    Map<String, Object> responseMap;
-    Collection<String> shardingKeys;
+  public Response getShardingKeys(@QueryParam("realm") String realm,
+      @QueryParam("prefix") String prefix) {
     try {
-      // If realm is not set in query param, the endpoint is: "/sharding-keys"
-      // to get all sharding keys in a namespace.
-      if (realm == null) {
-        shardingKeys = _metadataStoreDirectory.getAllShardingKeys(_namespace);
-        // To avoid allocating unnecessary resource, limit the map's capacity only for
-        // SHARDING_KEYS.
-        responseMap = new HashMap<>(1);
+      if (realm == null && prefix == null) {
+        // For endpoint: "/sharding-keys" to get all sharding keys in a namespace.
+        return getAllShardingKeys();
+      } else if (prefix == null) {
+        // For endpoint: "/sharding-keys?realm={realm}"
+        return getAllShardingKeysInRealm(realm);
+      } else if (realm == null) {
+        // For endpoint: "/sharding-keys?prefix={prefix}"
+        return getAllShardingKeysUnderPath(prefix);
       } else {
-        // For endpoint: "/sharding-keys?realm={realmName}"
-        shardingKeys = _metadataStoreDirectory.getAllShardingKeysInRealm(_namespace, realm);
-        // To avoid allocating unnecessary resource, limit the map's capacity only for
-        // SHARDING_KEYS and SINGLE_METADATA_STORE_REALM.
-        responseMap = new HashMap<>(2);
-        responseMap.put(MetadataStoreRoutingConstants.SINGLE_METADATA_STORE_REALM, realm);
+        // For endpoint: "/sharding-keys?realm={realm}&prefix={prefix}"
+        return getRealmShardingKeysUnderPath(realm, prefix);
```

Comment:
This logic is very complex. Can we leverage the default value to simplify the logic?

Reply:
A new endpoint which is equivalent to `GET /sharding-keys?realm={realm}`  is added: `GET /metadata-store-realms/{realm}/sharding-keys`

Reply:
There is a good way to simplify the logic here utilizing the default values as @dasahcc has mentioned. If the prefix is not provided, consider assigning "/" to the prefix variable and use the same logic as the situation where prefix exists.

Reply:
> There is a good way to simplify the logic here utilizing the default values as @dasahcc has mentioned. If the prefix is not provided, consider assigning "/" to the prefix variable and use the same logic as the situation where prefix exists.

@NealSun96 I don't think adding "/" to the prefix var is a better idea because:
- The Json responses are different for each case here. If you assign "/" to a method, the response could have `{ "prefix": "/" }`
- I've tried to find a good default value for this prefix. If I assign "/" as a default value, I've no idea how to differentiate if prefix is not provide or it is "/".
- The code logic to generate response map is different.

If I don't understand your idea, feel free to clarify it. Thanks for the tip.

Reply:
When the prefix parameter is not provided, it's treated as the caller doesn't care about the prefix - this method is doing "return all keys regardless of any prefix". This behavior is functionally the same as if the caller provides "/" as the prefix. I think it's reasonable to use "/" as a default value.

Reply:
public static final String ROUTING_DATA_PATH = "/METADATA_STORE_ROUTING_DATA";

This should be the default value.